tests/unit/test_dimension_utilities.py

```diff
@@ -304,6 +320,165 @@ def test_prefetch_dimension_variables(self, mock_get_opendap_nc4):
+            3) True  - cell_alignment[edge] attribute exists and
+                       bounds variable does not exist.
+            4) False - cell_alignment[edge] attribute does not exist and
+                       bounds variable does not exist.
+
+        """
+        varinfo_bounds = VarInfoFromDmr(
+            'tests/data/ATL16_prefetch_bnds.dmr'
+        )
+
+        with self.subTest('Variable has cell alignment and bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_edge_has_bnds')))
+
+        with self.subTest('Variable has no cell alignment and has bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_no_edge_has_bnds')))
+
+        with self.subTest('Variable has cell alignment and no bounds'):
+            self.assertTrue(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_edge_no_bnds')))
+
+        with self.subTest('Variable has no cell alignment and no bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_no_edge_no_bnds')))
+
+    def test_get_bounds_array(self):
+        """ Ensure that the expected bounds array is created given
+            the input dimension variable values.
+
+        """
+        prefetch_dataset = Dataset('tests/data/ATL16_prefetch.nc4', 'r')
+        dimension_path = '/npolar_grid_lat'
+
+        expected_bounds_array = self.bounds_array
+
+        assert_array_equal(get_bounds_array(prefetch_dataset,
+                                            dimension_path),
+                           expected_bounds_array)
+
+    def test_write_bounds(self):
+        """ Ensure that bounds data array is written to the dimension
+            dataset, both when the dimension variable is in the root group
+            and in a nested group.
+
+        """
+        varinfo_prefetch = VarInfoFromDmr('tests/data/ATL16_prefetch_group.dmr')
+        prefetch_dataset = Dataset('tests/data/ATL16_prefetch_group.nc4', 'r+')
+
+        # Expected variable contents in file.
+        expected_bounds_data = self.bounds_array
+
+        with self.subTest('Dimension variable is in the root group'):
+            root_variable_full_path = '/npolar_grid_lat'
+            root_varinfo_variable = varinfo_prefetch.get_variable(
+                root_variable_full_path)
+            root_variable_name = 'npolar_grid_lat'
+            root_bounds_name = root_variable_name + '_bnds'
+
+            write_bounds(prefetch_dataset, root_varinfo_variable)
+
+            resulting_bounds_root_data = prefetch_dataset.variables[
+                root_bounds_name][:]
+            assert_array_equal(resulting_bounds_root_data,
+                               expected_bounds_data)
+            # Check that bounds variable exists in the root group.
+            self.assertTrue(prefetch_dataset.variables[root_variable_name])
```

Comment:
This is a nit, and you can ignore it, but it feels like maybe this assertion logically could go before you try and access the values in the bounds variables.

If the bounds variable doesn't exist, the code will likely throw some sort of exception on L430 (where you try and access the array values), which means you'll never actually run this assertion when it would fail.

Reply:
I can see that logically. If the bounds variable doesn't exist, there's no point in doing the rest of the tests. Also, should this be:

```
# Check that bounds variable exists in the root group.
self.assertTrue(prefetch_dataset.variables[root_bounds_name])
```

with `root_bounds_name` instead of `root_variable_name`? Might've just caught my own mistake here.

Reply:
Yeah, I have the right `nested_bounds_variable` in the nested group test, so that must have been a copy/paste error.

Reply:
Good catch - I also hadn't spotted that, but I think you are right. Nice!